### ISSUE TITLE
fetch-fic freezes rather than do anything

### ISSUE THREAD

**Issue by alethiophile** [iarna/xenforo-to-epub#7]
When I try to run ff, it freezes. `htop` says it's spinning at 100% CPU. It shows no behavior at all, not even displaying the options help when run bare, just hangs taking up CPU cycles. Working with back versions, it seems this problem was introduced in 3.8.0.

**Comment by iarna** [comment]
It's been approximately forever since I did a release but since other folks are actually trying to use it, I'll do that, it ought to clear things up! (At least, I've had no problems with it…)

**Comment by iarna** [comment]
If you want to give it a go before I get a release pushed out, you can do that with:

`npx iarna/fetch-fic#latest` in place of `ff`. So like, downloading a fic would be…

`npx iarna/fetch-fic#latest get --scrape https://forums.spacebattles.com/posts/42459709` (which gets my most recent snippet)

Ah, `npx` is available if you have npm@5 I should add. It can be installed stand alone on older versions (`npm i -g npx`)

**Comment by alethiophile** [comment]
Trying that command, I get the same issue: npx installs the code, runs the
binary, and it jumps to 100% CPU and hangs without ever doing anything.

I have no particular expertise in debugging node stuff. I can try to run in a
different environment once I get a chance.

On Wed, Apr 18, 2018 at 11:05:36PM -0700, Rebecca Turner wrote:
> If you want to give it a go before I get a release pushed out, you can do that with:
> 
> `npx iarna/fetch-fic#latest` in place of `ff`. So like, downloading a fic would be…
> 
> `npx iarna/fetch-fic#latest get --scrape https://forums.spacebattles.com/posts/42459709` (which gets my most recent snippet)
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/iarna/fetch-fic/issues/7#issuecomment-382620918

-- 
Tom Hunt

**Comment by iarna** [comment]
@alethiophile What OS are you on? I've not tested this on Windows yet, but I probably should…

**Comment by alethiophile** [comment]
I'm on a stock Linux. I tested on Arch (node v9.8.0, npx 9.7.1) and also Fedora (node v9.4.0, npx 9.7.1). Both showed the same behavior of freezing in CPU-spin mode without doing anything, both using that command line (`npx iarna/fetch-fic#latest`). I tested with just the bare command, which if it was working properly should just print a usage message AFAICT.

Tomorrow I can try doing an install and getting an editable file to run tests with.

**Comment by iarna** [comment]
Huuh, ok, I'll give it try from Linux tomorrow.

**Comment by alethiophile** [comment]
Okay, so this is very weird.

1. It looks like the issue may not be specific to fetch-fic at all; I created a test file with the contents
   ```
   #!/usr/bin/env node --max_old_space_size=4096
   'use strict'

   console.log("at file top");
   process.exit(0);
   ```
   and this still exhibits the behavior when run as `./node_test.js`.

2. When run manually as `node node_test.js`, or even the full `/usr/bin/env node --max_old_space_size=4096 node_test.js`, it works as expected, printing and exiting.

3. When I run `node node_modules/fetch-fic/ff.js` manually, it works as well (at least to the extent of printing the usage message).

I have no idea what environmental difference there is between a shbang run and a manually-invoked run that would cause this.

**Comment by alethiophile** [comment]
On further investigation, if I remove the --max_old_space_size argument from the shebang, on either the test script or `ff.js` proper, it works.

I assume this is due to the [linux shebang argument issue](https://unix.stackexchange.com/questions/63979/shebang-line-with-usr-bin-env-command-argument-fails-on-linux), where on Linux only the first whitespace is split and the rest of the shebang is passed as a single argument. Using `strace`, it looks like `env` is interpreting "node --max_old_space_size=4096" as a variable-set argument (for an environment variable "node --max_old_space_size"), and executing the javascript file again without invoking `node` at all. Which then does the same thing again, so it's an infinite loop (luckily one only within the same process).

Is the --max_old_space_size argument strongly necessary?

**Comment by alethiophile** [comment]
If setting max_old_space_size is in fact necessary, [this page](https://github.com/npm/npm/issues/12238) seems to describe a way of doing so without the command-line argument (though it needs a variable set).

**Comment by iarna** [comment]
For some truly exceptionally large fic it is. I totally forgot about that Linux limitation!

As it happens, since that page was written, Node has introduced an environment variable you can set to pass command line args. It's still not super satisfying though. hmm…

**Comment by alethiophile** [comment]
There's also the "two-line shebang" (shell polyglot) from [here](https://stackoverflow.com/questions/20638520/appropriate-hashbang-for-node-js-scripts/28646724). Not sure how that will work on other systems than Linux.

**Comment by iarna** [comment]
The two-line thing is interesting and certainly would work on Mac and other BSDs, but is not compatible with [cmd-shim](https://www.npmjs.com/package/cmd-shim) (which is what npm uses to make shebangs work on windows). Hm, I think I will likely jus take it out and put it in the docs as a thing to put in your environment if you need it. There's only one fic I know of that that actually hits it anyway.

**Comment by alethiophile** [comment]
Sounds good.

Of note, on the same fic that ran out the stack space (#8), I also had to bump the max_old_space_size parameter to 8192 during generate. That particular one is very long and has short average chapters, so probably out near the edge of stress testing.

**Comment by iarna** [comment]
The size argument is necessary to produce epub of some larger fic, at least until I spend some time debugging memory usage. (Which probably means rewriting the entire epub generation flow.)

I'll take it out of the default command line and keep it as a local floating patch. (I have a few others like that.)

**Comment by iarna** [comment]
Yeah, epub generation is where it happens and fixing that probably means redoing the whole epub flow. I went to a lot of trouble to make it work as Node streams end-to-end, but in practice it ends up holding a lot more in memory than we really want, in particular because a streamed approach means that some things need to be buffered because epub has rules about zip file ordering. If I move to a random-access model it should let that be both simpler and less memory hungry. It'd also make the other output modes a lot simpler. Unfortunately it means rewriting the entire output subsystem.
